include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2023 Cirrus Logic, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+
+/**
+ * @brief Charger Interface
+ * @defgroup charger_interface Charger Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+/**
+ * @brief Runtime Dynamic Battery Parameters
+ */
+enum charger_props {
+	/** Indicates if external supply is present for the charger */
+	CHARGER_PROP_ONLINE = 0,
+	/** Reports whether or not a battery is present */
+	CHARGER_PROP_PRESENT,
+	/** Represents the charging status of the charger */
+	CHARGER_PROP_STATUS,
+	/** Reserved to demark end of common charger properties */
+	CHARGER_PROP_COMMON_COUNT,
+	/**
+	 * Reserved to demark downstream custom properties - use this value as the actual value may
+	 * change over future versions of this API
+	 */
+	CHARGER_PROP_CUSTOM_BEGIN = CHARGER_PROP_COMMON_COUNT + 1,
+	/** Reserved to demark end of valid enum properties */
+	CHARGER_PROP_MAX = UINT16_MAX,
+};
+
+/**
+ * @brief External supply states
+ */
+enum charger_online {
+	/** External supply not presented */
+	CHARGER_ONLINE_OFFLINE = 0,
+	/** External supply is present and of fixed output */
+	CHARGER_ONLINE_FIXED,
+	/** External supply is present and of programmable output*/
+	CHARGER_ONLINE_PROGRAMMABLE,
+};
+
+/**
+ * @brief Charging states
+ */
+enum charger_status {
+	/** Charging device state is unknown */
+	CHARGER_STATUS_UNKNOWN = 0,
+	/** Charging device is charging a battery */
+	CHARGER_STATUS_CHARGING,
+	/** Charging device is not able to charge a battery */
+	CHARGER_STATUS_DISCHARGING,
+	/** Charging device is not charging a battery */
+	CHARGER_STATUS_NOT_CHARGING,
+	/** The battery is full and the charging device will not attempt charging */
+	CHARGER_STATUS_FULL,
+};
+
+/**
+ * @typedef charger_property_t
+ * @brief A charger property's identifier
+ *
+ * See charger_props for a list of identifiers
+ */
+typedef uint16_t charger_property_t;
+
+/**
+ * @brief Representation of a charger property
+ *
+ * The property has a property type that identifies the property to get,
+ * an error associated with fetching the property, and the fetched value
+ * of the property.
+ */
+struct charger_get_property {
+	/** Battery charger property to get */
+	charger_property_t property_type;
+	/** Negative error status set by callee e.g. -ENOTSUP for an unsupported property */
+	int err;
+	/** Property field for getting */
+	union {
+		enum charger_online online;
+		bool present;
+		enum charger_status status;
+	} value;
```

Comment:
The union just has a limited set, but what about custom properties? What if a new custom property doesn't fit any? It could be a blocker for oot drivers. Maybe it should be a typedeffed integer type?

Reply:
ack, I'll implement the typedef